admin/javascript/src/components/form-builder/index.js

```diff
@@ -0,0 +1,229 @@
+   *
+   * @param boolean schema - If form schema data should be returned in the response.
+   * @param boolean state - If form state data should be returned in the response.
+   *
+   * @return object - Promise from the AJAX request.
+   */
+  fetch(schema = true, state = false) {
+    const headerValues = [];
+
+    if (this.isFetching === true) {
+      return this.formSchemaPromise;
+    }
+
+    if (schema === true) {
+      headerValues.push('schema');
+    }
+
+    if (state === true) {
+      headerValues.push('state');
+    }
+
+    this.formSchemaPromise = fetch(this.props.schemaUrl, {
+      headers: { 'X-FormSchema-Request': headerValues.join() },
+      credentials: 'same-origin',
+    })
+      .then(response => response.json())
+      .then(json => {
+        this.isFetching = false;
+        this.props.actions.setSchema(json);
+      });
+
+    this.isFetching = true;
+
+    return this.formSchemaPromise;
+  }
+
+  /**
+   * Maps a list of schema fields to their React Component.
+   * Only top level form fields are handled here, composite fields (TabSets etc),
+   * are responsible for mapping and rendering their children.
+   *
+   * @param array fields
+   *
+   * @return array
+   */
+  mapFieldsToComponents(fields) {
+    return fields.map((field, i) => {
+      const Component = field.component !== null
+        ? fakeInjector.getComponentByName(field.component)
+        : fakeInjector.getComponentByDataType(field.type);
+
+      if (Component === null) {
+        return null;
+      }
+
+      // Props which every form field receives.
+      const props = {
+        attributes: field.attributes,
+        data: field.data,
+        description: field.description,
+        extraClass: field.extraClass,
+        fields: field.children,
+        id: field.id,
+        name: field.name,
+      };
+
+      // Structural fields (like TabSets) are not posted back to
+      // the server and don't receive some props.
+      if (field.type !== 'Structural') {
+        props.rightTitle = field.rightTitle;
+        props.leftTitle = field.leftTitle;
+        props.readOnly = field.readOnly;
+        props.disabled = field.disabled;
+        props.customValidationMessage = field.customValidationMessage;
+      }
+
+      // Dropdown and Radio fields need some options...
+      if (field.type === 'MultiSelect' || field.type === 'SingleSelect') {
+        props.source = field.source;
+      }
+
+      return <Component key={i} {...props} />;
+    });
+  }
+
+  render() {
+    const schema = this.props.schemas[this.props.schemaUrl];
+
+    // If the response from fetching the initial data
+    // hasn't come back yet, don't render anything.
+    if (!schema) {
+      return null;
+    }
+
+    const formProps = {
+      actions: schema.schema.actions,
+      attributes: schema.schema.attributes,
+      data: schema.schema.data,
+      fields: schema.schema.fields,
+      mapFieldsToComponents: this.mapFieldsToComponents,
```

Comment:
The repetitive naming here is a bit awkward. Can we name `const schema` something else for clarity? It's weird that `schema` is a property of a schema.